backend/tracim_backend/lib/event.py

```diff
@@ -0,0 +1,287 @@
+from tracim_backend.models.auth import Profile
+from tracim_backend.models.auth import User
+from tracim_backend.models.data import Content
+from tracim_backend.models.data import UserRoleInWorkspace
+from tracim_backend.models.data import Workspace
+from tracim_backend.models.data import WorkspaceRoles
+from tracim_backend.models.event import EntityType
+from tracim_backend.models.event import Event
+from tracim_backend.models.event import Message
+from tracim_backend.models.event import OperationType
+from tracim_backend.models.event import ReadStatus
+from tracim_backend.models.tracim_session import TracimSession
+from tracim_backend.views.core_api.schemas import ContentDigestSchema
+from tracim_backend.views.core_api.schemas import EventSchema
+from tracim_backend.views.core_api.schemas import UserDigestSchema
+from tracim_backend.views.core_api.schemas import WorkspaceDigestSchema
+
+_USER_FIELD = "user"
+_AUTHOR_FIELD = "author"
+_WORKSPACE_FIELD = "workspace"
+_CONTENT_FIELD = "content"
+_ROLE_FIELD = "role"
+
+
+RQ_QUEUE_NAME = "event"
+
+
+class EventApi:
+    """Api to query event & messages"""
+
+    def __init__(self, current_user: User, session: TracimSession, config: CFG) -> None:
+        self._current_user = current_user
+        self._session = session
+        self._config = config
+
+    def get_messages_for_user(self, user_id: int, read_status: ReadStatus) -> typing.List[Message]:
+        query = (
+            self._session.query(Message)
+            .filter(Message.receiver_id == user_id)
+            .options(joinedload(Message.event))
+        )
+        if read_status == ReadStatus.READ:
+            query = query.filter(Message.read != null())
+        elif read_status == ReadStatus.UNREAD:
+            query = query.filter(Message.read == null())
+        return query.all()
+
+
+class EventBuilder:
+    """Create Event objects from the database crud hooks."""
+
+    _user_schema = UserDigestSchema()
+    _workspace_schema = WorkspaceDigestSchema()
+    _content_schema = ContentDigestSchema()
+    _event_schema = EventSchema()
+
+    def __init__(self, config: CFG) -> None:
+        self._config = config
+        self._current_user = None  # type: typing.Optional[User]
+
+    @hookimpl
+    def on_current_user_set(self, user: User) -> None:
+        self._current_user = user
+
+    @hookimpl
+    def on_request_session_created(self, request: TracimRequest, session: TracimSession) -> None:
+        event.listen(session, "pending_to_persistent", self._publish_event)
+
+    @hookimpl
+    def on_request_finished(self) -> None:
+        self._current_user = None
+
+    # User events
+    @hookimpl
+    def on_user_created(self, user: User, db_session: TracimSession) -> None:
+        self._create_user_event(OperationType.CREATED, user, db_session)
+
+    @hookimpl
+    def on_user_modified(self, user: User, db_session: TracimSession) -> None:
+        if user.is_deleted:
```

Comment:
Are we sure there aren't edge cases where a deleted user is modified, making this function behave erratically (if I understand its logic correctly)? 

Same remark for all these functions using `something.is_deleted`

Reply:
@grignards , is there any way to have some kind of diff between previews and new User at this level ?

Reply:
Yes it is possible through [the attribute state](https://docs.sqlalchemy.org/en/13/orm/internals.html#sqlalchemy.orm.state.AttributeState.history)

Reply:
Attribute state works as expected, I modified my code to ensure behavior (plus a unit test for the corner case of is_deleted==True then modified).